File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
	if len(prices) == 0 {
		return 0
	}

	maxProfit := 0
	minPrice := prices[0]

	for _, price := range prices {
		// If we find any price which is lower than the current minPrice
		// update the minPrice
		if price < minPrice {
			minPrice = price
		} else if profit := price - minPrice; profit > maxProfit {
			// If diff of current stock with minPrice is greater
			// update the profit
			maxProfit = profit
		}
	}

	return maxProfit
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) {
            return 0;
        }

        int maxProfit = 0;
        int minPrice = prices[0];

        for (int i = 0; i < prices.length; i++) {
            int profit = prices[i] - minPrice;

            if (prices[i] < minPrice) {
                minPrice = prices[i];
            } else if (profit > maxProfit) {
                maxProfit = prices[i] - minPrice;
            }
        }

        return maxProfit;
    }
}
